### PR TITLE
Remove operator packages from FBC index image

### PR DESCRIPTION
**Pull request by chandwanitulsi**
This commit is for removing the specified operator from
FBC index image. The image is identified with the label
"operators.operatorframework.io.index.configs.v1".
If the index image is FBC, the operator is removed
by using opm registry commands and then index is migrated
to FBC format, generating the dockerfile and building the
resultant image in the end.

Refers to CLOUDDST-8616

### CODE REVIEW

File: iib/workers/tasks/build.py [hunk 1]
```diff
@@ -907,17 +912,18 @@ def handle_rm_request(
     with tempfile.TemporaryDirectory(prefix='iib-') as temp_dir:
         with set_registry_token(overwrite_from_index_token, from_index_resolved):
             if is_image_fbc(from_index_resolved):
-                err_msg = 'File-Based catalog image type is not supported yet.'
-                log.error(err_msg)
-                raise IIBError(err_msg)
+                opm_registry_rm_fbc(
```
Comment (lipoja) [inline iib/workers/tasks/build.py:915]: could you please log info message that we are working with FBC image.

File: iib/workers/tasks/opm_operations.py [hunk 2]
```diff
@@ -389,3 +389,57 @@ def opm_registry_add_fbc(
...
+    cmd = [
+        'opm',
+        'registry',
+        'rm',
+        '--database',
+        index_db_path,
+        '--packages',
+        ','.join(operators),
+    ]
+    run_cmd(cmd, {'cwd': base_dir}, exc_msg='Failed to add the bundles to the index image')
```
Comment (lipoja) [inline iib/workers/tasks/opm_operations.py:414]: `Failed to add` :) but we are removing :)

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 3]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
     cmd = ['opm', 'registry', 'serve', '-p', '50051', '-d', '/tmp/dummy.db', '-t', '/dev/null']
     opm_operations._serve_cmd_at_port(" ".join(cmd), '/tmp', 50051, 5, 3)
     assert mock_run_cmd.call_count == 7
+
+
+@mock.patch('iib.workers.tasks.opm_operations.shutil.rmtree')
+@mock.patch('iib.workers.tasks.utils.run_cmd')
+def test_opm_migrate(
```
Comment (lipoja) [inline tests/test_workers/test_tasks/test_opm_operations.py:163]: This belongs to different commit - I guess it will be cleaned once properly we rebase our code.

**Comment by lipoja** [comment]
@chandwanitulsi Looks OK, I guess deeper review will be done once we rebase the code.

File: iib/workers/tasks/opm_operations.py (hunk 2 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/opm_operations.py:414]: copy paste error :P

File: iib/workers/tasks/build.py (hunk 1 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build.py:915]: I have added log inside opm_registry_rm_fbc(), would you rather have it added here instead?

File: tests/test_workers/test_tasks/test_opm_operations.py (hunk 3 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_opm_operations.py:163]: yup, that's right

File: tests/test_workers/test_tasks/test_build.py [hunk 4]
```diff
@@ -1004,18 +1004,80 @@ def test_handle_rm_request(
...
+        'binary_image': 'binary-image:latest',
+        'binary_image_resolved': 'binary-image@sha256:abcdef',
+        'from_index_resolved': 'from-index@sha256:bcdefg',
+        'ocp_version': 'v4.6',
+        'distribution_scope': 'PROD',
+    }
+    mock_ghid.return_value = "/tmp/xyz/database/index.db"
+    mock_ogd.return_value = "/tmp/xyz/catalog.Dockerfile"
+    mock_om.return_value = "/tmp/xyz/catalog"
+    mock_orrf.return_value = "/catalog.Dockerfile"
```
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build.py:1054]: I logged the return response from opm_generate_dockerfile() which is also returned by and that is something like /tmp/iib-s51a41v3/index.Dockerfile. So /catalog.Dockerfile is not correct but its not relative as well

**Comment by chandwanitulsi** [comment]
@release-engineering/exd-guild-hello-operator please review

File: iib/workers/tasks/build.py [hunk 5]
```diff
@@ -907,17 +912,21 @@ def handle_rm_request(
...
-        _opm_index_rm(
-            temp_dir,
-            operators,
-            prebuild_info['binary_image'],
-            from_index_resolved,
-            overwrite_from_index_token,
-        )
+            else:
+                _opm_index_rm(
+                    temp_dir,
```
Comment (pbortlov) [inline iib/workers/tasks/build.py:924]: Please add value=value mapping

File: iib/workers/tasks/opm_operations.py [hunk 6]
```diff
@@ -389,3 +389,57 @@ def opm_registry_add_fbc(
         binary_image=binary_image,
         dockerfile_name='index.Dockerfile',
     )
+
+
+def _opm_registry_rm(index_db_path, operators, base_dir):
+    from iib.workers.tasks.utils import run_cmd
```
Comment (pbortlov) [inline iib/workers/tasks/opm_operations.py:395]: could you add it bellow the docstring?

File: tests/test_workers/test_tasks/test_build.py [hunk 7]
```diff
@@ -1004,18 +1004,80 @@ def test_handle_rm_request(
...
+@mock.patch('iib.workers.tasks.utils.set_request_state')
+@mock.patch('iib.workers.tasks.build.set_request_state')
+@mock.patch('iib.workers.tasks.build._create_and_push_manifest_list')
+@mock.patch('iib.workers.tasks.build._update_index_image_pull_spec')
+@mock.patch('iib.workers.tasks.build._add_label_to_index')
 @mock.patch('iib.workers.tasks.build.is_image_fbc')
-def test_handle_rm_request_raises(mock_iifbc, mock_runcmd, mock_c):
+@mock.patch('iib.workers.tasks.opm_operations.get_hidden_index_database')
+@mock.patch('iib.workers.tasks.opm_operations.opm_migrate')
+@mock.patch('iib.workers.tasks.opm_operations.omp_generate_dockerfile')
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_build.py:1022]: typo OMP

File: tests/test_workers/test_tasks/test_build.py [hunk 8]
```diff
@@ -1004,18 +1004,80 @@ def test_handle_rm_request(
 
 
 @mock.patch('iib.workers.tasks.build._cleanup')
-@mock.patch('iib.workers.tasks.build.run_cmd')
+@mock.patch('iib.workers.tasks.build.prepare_request_for_build')
+@mock.patch('iib.workers.tasks.build._update_index_image_build_state')
+@mock.patch('iib.workers.tasks.build.opm_registry_rm_fbc')
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_build.py:1009]: isn't this function in opm_operations?

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
+import os.path
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:2]: could you group `import` an `from library import function` together?

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 10]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+        ['opm', 'migrate', index_db_file, os.path.join(tmpdir, 'catalog')],
+        {'cwd': tmpdir},
+        exc_msg='Failed to migrate index.db to file-based catalog',
+    )
+
+
+@pytest.mark.parametrize("set_index_db_file", (False, True))
+@pytest.mark.parametrize("dockerfile", (None, 'index.Dockerfile'))
+@mock.patch('iib.workers.tasks.utils.run_cmd')
+def test_omp_generate_dockerfile(mock_run_cmd, tmpdir, dockerfile, set_index_db_file):
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:181]: opm?

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 11]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+    index_db_file = os.path.join(tmpdir, 'database/index.db') if set_index_db_file else None
+    fbc_dir = os.path.join(tmpdir, 'catalogs')
+
+    def crate_dockerfile(*args, **kwargs):
+        with open(os.path.join(tmpdir, 'catalogs.Dockerfile'), 'a'):
+            pass
+
+    mock_run_cmd.side_effect = crate_dockerfile
+
+    opm_operations.omp_generate_dockerfile(
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:191]: opm?

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 12]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+    else:
+        with open(df_path, 'r') as f:
+            assert all(
+                line.find('/var/lib/iib/_hidden/do.not.edit.db') == -1 for line in f.readlines()
+            )
+
+
+@pytest.mark.parametrize("set_index_db_file", (False, True))
+@mock.patch('iib.workers.tasks.utils.run_cmd')
+def test_omp_generate_dockerfile_no_dockerfile(mock_run_cmd, tmpdir, set_index_db_file):
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:218]: opm

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 13]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+    index_db_file = os.path.join(tmpdir, 'database/index.db') if set_index_db_file else None
+    fbc_dir = os.path.join(tmpdir, 'catalogs')
+    df_path = os.path.join(tmpdir, f"{os.path.basename(fbc_dir)}.Dockerfile")
+
+    with pytest.raises(IIBError, match=f"Can not find generated Dockerfile at {df_path}"):
+        opm_operations.omp_generate_dockerfile(fbc_dir, tmpdir, index_db_file)
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:224]: opm

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 14]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+    assert "--enable-alpha" in opm_args
+
+    mock_srt.assert_called_once_with('user:pass', from_index)
+
+
+@pytest.mark.parametrize('from_index', (None, 'some_index:latest'))
+@pytest.mark.parametrize('bundles', (['bundle:1.2', 'bundle:1.3'], []))
+@pytest.mark.parametrize('overwrite_csv', (True, False))
+@pytest.mark.parametrize('container_tool', (None, 'podwoman'))
+@mock.patch('iib.workers.tasks.opm_operations.omp_generate_dockerfile')
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:277]: opm

File: tests/test_workers/test_tasks/test_opm_operations.py [hunk 15]
```diff
@@ -155,3 +156,214 @@ def test_serve_cmd_at_port_delayed_initialize(
...
+        fbc_dir=fbc_dir,
+        base_dir=tmpdir,
+        index_db=index_db_file,
+        binary_image="some:image",
+        dockerfile_name='index.Dockerfile',
+    )
+
+
+@pytest.mark.parametrize('operators', (['abc-operator', 'xyz-operator'], []))
+@mock.patch('iib.workers.tasks.opm_operations.omp_generate_dockerfile')
```
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_opm_operations.py:328]: opm

**Review by pbortlov — CHANGES_REQUESTED** [review]

File: tests/test_workers/test_tasks/test_build.py (hunk 7 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build.py:1022]: This PR is not based on latest [add pr](https://github.com/release-engineering/iib/pull/334/) and earlier it had some typos, which got carry forward here. Once I rebase this PR, all this would be fixed

File: tests/test_workers/test_tasks/test_build.py (hunk 8 above)
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build.py:1009]: I had to patch the where the function is called as well

File: iib/workers/tasks/build.py (hunk 1 above)
Comment (lipoja) [reply iib/workers/tasks/build.py:915]: I know that it was requested by Tim. I thought that we might do this (or similar):
```
is_fbc = is_image_fbc(from_index_resolved)
if is_fbc:
    log.info("Processing File-Based Catalog image")
```
in every `handle_` function. So it is consistent across all API endpoints (ideally the same message, so it is easy to find/grep)
Does it sounds reasonable?

File: tests/test_workers/test_tasks/test_opm_operations.py (hunk 14 above)
Comment (lipoja) [reply tests/test_workers/test_tasks/test_opm_operations.py:277]: rebase should fix it - this is not related to this PR. Resolving conversation

File: iib/workers/tasks/build.py (hunk 1 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build.py:938]: Added log here as well :)

File: iib/workers/tasks/opm_operations.py [hunk 16]
```diff
@@ -382,3 +382,58 @@ def opm_registry_add_fbc(
         binary_image=binary_image,
         dockerfile_name='index.Dockerfile',
     )
+
+
+def _opm_registry_rm(index_db_path, operators, base_dir):
```
[the remark this answers is not in the document]
Comment (pbortlov) [reply iib/workers/tasks/opm_operations.py:387]: but then it has to be changed in create_empty_index too. Let's keep it this way.

**Review by pbortlov — CHANGES_REQUESTED** [review]
tests are failing

**Comment by chandwanitulsi** [comment]
> 

Yup it was failing in flake, fixed it :)

File: iib/workers/tasks/opm_operations.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/workers/tasks/opm_operations.py:387]: opm_registry_rm was newly created here and it should be used within this file

[the remark this answers is not in the document]
Comment (pbortlov) [reply iib/workers/tasks/opm_operations.py:387]: isn't it a duplicate of `_opm_index_rm` from build.py?

[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/opm_operations.py:387]: `_opm_index_rm ` works with index image but `_opm_registry_rm` works directly with `index.db` file. They are not duplicates.

**Review by JAVGan — APPROVED** [review]

File: iib/workers/tasks/opm_operations.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (JAVGan) [reply iib/workers/tasks/opm_operations.py:387]: Isn't it possible to make `_opm_index_rm` more generic to cover both cases instead?

**Comment by chandwanitulsi** [comment]
> Everything LGTM but please consider the idea of generic `_opm_index_rm`.
> 
> Because with this PR and #340 new functions are being created almost as the same as we had previously for non FBC, I know they're different but spreading similar code across the project might lead to future development confusion.

They are actually not similar in code, also in future we will have FBC based indexes only and all the operations will happen with opm registry commands so I think it will be easier to clean opm_index-* code

File: iib/workers/tasks/opm_operations.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/workers/tasks/opm_operations.py:387]: I think we wanted to have clear separation in index and registry commands

[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/opm_operations.py:387]: yes we want to have it separate.

File: tests/test_workers/test_tasks/test_build.py [hunk 17]
```diff
@@ -1019,18 +1019,80 @@ def test_handle_rm_request(
...
+@mock.patch('iib.workers.tasks.build._build_image')
+@mock.patch('iib.workers.tasks.build._push_image')
+@mock.patch('iib.workers.tasks.build._verify_index_image')
+@mock.patch('iib.workers.tasks.utils.set_request_state')
+@mock.patch('iib.workers.tasks.build.set_request_state')
+@mock.patch('iib.workers.tasks.build._create_and_push_manifest_list')
+@mock.patch('iib.workers.tasks.build._update_index_image_pull_spec')
+@mock.patch('iib.workers.tasks.build._add_label_to_index')
 @mock.patch('iib.workers.tasks.build.is_image_fbc')
-def test_handle_rm_request_raises(mock_iifbc, mock_runcmd, mock_c):
```
Comment (lipoja) [inline tests/test_workers/test_tasks/test_build.py:1024]: Was this test removed/replaced?

Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_build.py:1024]: The test was checking if error is raised when FBC index is provided in from_index,but that was chnanged, so I changed the test

Comment (lipoja) [reply tests/test_workers/test_tasks/test_build.py:1024]: Thanks I did not noticed that.

**Comment by lipoja** [comment]
@chandwanitulsi rebase please..

**Review by JAVGan — APPROVED** [review]

**Review by pbortlov — APPROVED** [review]
